Index: polkit-qt-1/agent/polkitqt1-agent-session.cpp
===================================================================
--- polkit-qt-1.orig/agent/polkitqt1-agent-session.cpp
+++ polkit-qt-1/agent/polkitqt1-agent-session.cpp
@@ -150,6 +150,12 @@ void AsyncResult::setCompleted()
     d->result = nullptr;
 }
 
+void AsyncResult::setCancel(const QString &text)
+{
+    Q_ASSERT(d->result);
+    g_simple_async_result_set_error(d->result, POLKIT_ERROR, POLKIT_ERROR_CANCELLED, "%s", text.toUtf8().data());
+}
+
 void AsyncResult::setError(const QString &text)
 {
     Q_ASSERT(d->result);
Index: polkit-qt-1/agent/polkitqt1-agent-session.h
===================================================================
--- polkit-qt-1.orig/agent/polkitqt1-agent-session.h
+++ polkit-qt-1/agent/polkitqt1-agent-session.h
@@ -45,6 +45,11 @@ public:
     void setCompleted();
 
     /**
+     * \brief Mark the action that is tied to this result as cancelled.
+     */
+    void setCancel(const QString &text);
+
+    /**
      * \brief Sets an error for the asynchronous result.
      * Method complete() must be called anyway.
      *
